pkg/storage/rulestore/bucket_client.go

```diff
@@ -0,0 +1,308 @@
+package rulestore
```

Comment:
This belongs under ruler package, perhaps `pkg/ruler/rules/rulestore`? (which is honestly pretty bad, but local and objectclient live there too). 

Instead of rulestore, we could call it `bucketstore`, since `local` and `objectclient` are also called based on implementation.

Reply:
I would prefer to move the functionality from that package here once the old `objectclient` is deprecated. I think it would make sense to centralize packages that revolve around object storage under the `storage/` package prefix. As for the relevant protos from that package, I think a new `rulestorepb` package could be created either under `pkg/storage/rulestore` or `pkg/ruler/rulestorepb`.

Reply:
> I think it would make sense to centralize packages that revolve around object storage client under the `storage/` package prefix.

This is a wrong approach to modularization. Implementation detail (using bucket client) is not main categorization here – but the fact that it is part of ruler component is.

> As for the relevant protos from that package, I think the rulestore proto definitions can be moved to a `rulestorepb` package under `pkg/ruler/rulestorepb`.

I don't think packages under `pkg/storage` should depend on packages under `pkg/ruler`.

Reply:
> This is a wrong approach to modularization. Implementation detail (using bucket client) is not main categorization here – but the fact that it is part of ruler component is.

Fair enough. It should be straightforward to move this to the `pkg/ruler/rulestore`.

Reply:
Done